### DISCUSSION TITLE
How to call methods with Type arguments?

### DISCUSSION THREAD

**Discussion by handerss-spotfire** [ClearFoundry/ClearScript#506] [Q&A] [answered]
If I try call a method which takes a `Type` as its first argument I get the following error:

```
Error: The best overloaded method match for 'ClearScriptPlaygroud.ScriptObject.DoSomethingWithType(System.Type)' has some invalid arguments
```

I guess this happens since ClearScript thinks I'm trying to call a generic method. What is the way to call these types of methods?

Repro:

```csharp
namespace ClearScriptPlaygroud;

public class ScriptObject
{
    public void DoSomethingWithType(Type aType) { }
}

internal class Program
{
    static void Main(string[] args)
    {
        using var engine = new V8ScriptEngine();

        engine.AddHostObject("host", new HostFunctions());
        var typeCollection = new HostTypeCollection("mscorlib");
        engine.Script["System"] = typeCollection.GetNamespaceNode("System");
        var obj = new ScriptObject();
        engine.AddHostObject("obj", obj);

        engine.Execute(@"obj.DoSomethingWithType(System.String.Type);");
    }
}
```

**Comment by ClearScriptLib** [comment, marked as answer]
Hi @handerss-tibco,

> I guess this happens since ClearScript thinks I'm trying to call a generic method.

Not quite. Consider this line in your code:

```C#
engine.Execute(@"obj.DoSomethingWithType(System.String.Type);");
```

.NET's [`String`](https://learn.microsoft.com/en-us/dotnet/api/system.string) class has no static property named "Type", so the JavaScript expression `System.String.Type` evaluates to `undefined`, which isn't a valid argument for `DoSomethingWithType`.

> What is the way to call these types of methods?

In the JavaScript environment, `System.String` is a _host type_; it isn't a [`Type`](https://learn.microsoft.com/en-us/dotnet/api/system.type) object. A host type is a special object that enables C#-like syntax and capabilities.

Suppose you wanted to call `DoSomethingWithType` in C#. The following wouldn't work:

```C#
obj.DoSomethingWithType(string);
```

Instead, you'd use the [`typeof`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#typeof-operator) operator to obtain a `Type` object:

```C#
obj.DoSomethingWithType(typeof(string));
```

In JavaScript, you can use ClearScript's [`typeOf`](https://clearscript.clearfoundry.net/Reference/html/M_Microsoft_ClearScript_HostFunctions_typeOf.htm) method to do the same thing:

```C#
engine.AllowReflection = true;
engine.Execute(@"obj.DoSomethingWithType(host.typeOf(System.String));");
```

Good luck!

**Reply by handerss-spotfire** [reply]
Thanks for the answer! Forgot to remove the ".Type" part when copy-pasting code, nonetheless you answered my question.

`typeOf` unfortunately requires `AllowReflection` which may be problematic in our use case.

**Comment by ClearScriptLib** [comment]
>`typeOf` unfortunately requires `AllowReflection` which may be problematic in our use case.

Ah. In that case, you could expose the required `Type` objects directly, provide an alternative for [`GetType`](https://learn.microsoft.com/en-us/dotnet/api/system.object.gettype), etc.

Cheers!
